stark/src/merkle_tree/index.rs

```diff
@@ -0,0 +1,196 @@
+    }
+
+    pub fn depth(self) -> usize {
+        let next_layer = (self.0 + 1).next_power_of_two();
+        // Usize should always be able to hold it's number of bits
+        let next_depth = usize::try_from(next_layer.trailing_zeros()).unwrap();
+        next_depth - 1
+    }
+
+    pub fn offset(self) -> usize {
+        self.0 - (1_usize << self.depth())
+    }
+
+    pub fn is_root(self) -> bool {
+        self.0 == 1
+    }
+
+    pub fn is_left(self) -> bool {
+        self.0 % 2 == 0
+    }
+
+    pub fn is_right(self) -> bool {
+        self.0 != 1 && self.0 % 2 == 1
+    }
+
+    pub fn is_left_most(self) -> bool {
+        self.0.is_power_of_two()
+    }
+
+    pub fn is_right_most(self) -> bool {
+        (self.0 + 1).is_power_of_two()
+    }
+
+    pub fn parent(self) -> Option<Self> {
+        if self.is_root() {
+            None
+        } else {
+            Some(Self(self.0 >> 1))
+        }
+    }
+
+    pub fn sibling(self) -> Option<Self> {
+        if self.is_root() {
+            None
+        } else {
+            Some(Self(self.0 ^ 1))
+        }
+    }
+
+    pub fn left_neighbor(self) -> Option<Self> {
+        if self.is_left_most() {
+            None
+        } else {
+            Some(Self(self.0 - 1))
+        }
+    }
+
+    pub fn right_neighbor(self) -> Option<Self> {
+        if self.is_right_most() {
+            None
+        } else {
+            Some(Self(self.0 + 1))
+        }
+    }
+
+    pub fn left_child(self) -> Self {
+        Self(self.0 << 1)
+    }
+
+    pub fn right_child(self) -> Self {
+        Self((self.0 << 1) | 1)
+    }
+
+    pub fn last_common_ancestor(self, other: Self) -> Self {
+        // Align their first bits all the way to the left
+        let a = self.0 << self.0.leading_zeros();
+        let b = other.0 << other.0.leading_zeros();
+
+        // Extract the longest common prefix
+        let prefix_length = (a ^ b).leading_zeros();
+        let prefix = a >> (0_usize.count_zeros() - prefix_length);
+        Self(prefix)
+    }
+}
+
+#[cfg(feature = "std")]
+impl fmt::Debug for Index {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Index({:}, {:})", self.depth(), self.offset())
+    }
+}
+
+#[cfg(any(test, feature = "quickcheck"))]
+use quickcheck::{Arbitrary, Gen};
+
+#[cfg(any(test, feature = "quickcheck"))]
+impl Arbitrary for Index {
+    fn arbitrary<G: Gen>(g: &mut G) -> Self {
+        Self(usize::arbitrary(g) + 1)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use quickcheck_macros::quickcheck;
+
+    #[quickcheck]
+    pub fn test_depth_offset_roundtrip(depth: usize, offset: usize) -> bool {
+        let depth = depth % Index::max_depth();
+        let offset = offset % Index::nodes_at_depth(depth);
+        let index = Index::from_depth_offset(depth, offset).unwrap();
+        index.depth() == depth && index.offset() == offset
+    }
+
+    #[quickcheck]
+    pub fn test_children(parent: Index) -> bool {
```

Comment:
Can you make this 
```
left.is_left() && right.is_right() && ...
```
?

Reply:
I had that originally and didn't like it. When one of the subtests fails the backtrace will not tell you which one.